presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownSpatialProjections.java

```diff
@@ -0,0 +1,207 @@
+        }
+
+        JoinNode joinNode = captures.get(JOIN);
+
+        Expression filter = node.getPredicate();
+        List<FunctionCall> spatialFunctions = extractSupportedSpatialFunctions(filter);
+        for (FunctionCall spatialFunction : spatialFunctions) {
+            List<Expression> arguments = spatialFunction.getArguments();
+            Verify.verify(arguments.size() == 2);
+
+            Expression firstArgument = arguments.get(0);
+            Expression secondArgument = arguments.get(1);
+
+            Set<String> firstSymbols = getSymbolNames(extractUnique(firstArgument));
+            Set<String> secondSymbols = getSymbolNames(extractUnique(secondArgument));
+
+            if (firstSymbols.isEmpty() || secondSymbols.isEmpty()) {
+                return Result.empty();
+            }
+
+            Optional<Symbol> newFirstSymbol = newSymbol(context, firstArgument);
+            Optional<Symbol> newSecondSymbol = newSymbol(context, secondArgument);
+
+            Expression newFirstArgument = toExpression(newFirstSymbol, firstArgument);
+            Expression newSecondArgument = toExpression(newSecondSymbol, secondArgument);
+
+            Expression newSpatialFunction = new FunctionCall(spatialFunction.getName(), ImmutableList.of(newFirstArgument, newSecondArgument));
+            Expression newFilter = replaceExpression(filter, ImmutableMap.of(spatialFunction, newSpatialFunction));
+
+            PlanNode leftNode = joinNode.getLeft();
+            PlanNode rightNode = joinNode.getRight();
+
+            Set<String> leftSymbols = getSymbolNames(leftNode.getOutputSymbols());
+            Set<String> rightSymbols = getSymbolNames(rightNode.getOutputSymbols());
+
+            PlanNode newLeftNode;
+            PlanNode newRightNode;
+
+            if (leftSymbols.containsAll(firstSymbols)
+                    && containsNone(leftSymbols, secondSymbols)
+                    && rightSymbols.containsAll(secondSymbols)
+                    && containsNone(rightSymbols, firstSymbols)) {
+                newLeftNode = newFirstSymbol.map(symbol -> addProjection(context, leftNode, symbol, firstArgument)).orElse(leftNode);
+                newRightNode = newSecondSymbol.map(symbol -> addProjection(context, rightNode, symbol, secondArgument)).orElse(rightNode);
+            }
+            else if (leftSymbols.containsAll(secondSymbols)
+                    && containsNone(leftSymbols, firstSymbols)
+                    && rightSymbols.containsAll(firstSymbols)
+                    && containsNone(rightSymbols, secondSymbols)) {
+                newLeftNode = newSecondSymbol.map(symbol -> addProjection(context, leftNode, symbol, secondArgument)).orElse(leftNode);
+                newRightNode = newFirstSymbol.map(symbol -> addProjection(context, rightNode, symbol, firstArgument)).orElse(rightNode);
+            }
+            else {
+                continue;
+            }
+
+            return Result.ofPlanNode(new JoinNode(
```

Comment:
in case you always `continue` in 153 then this will make iterative optimizer to loop forever as it will be always able to transform the plan

Reply:
I don't understand. If we don't find a join condition that triggers a rewrite, the rule will return Result.empty();  on line 168 which will signal that the plan can't be transformed anymore. Am I missing something?

Reply:
ah right...  sorry about this. 

BTW. Why do you return from this rule after first function call rewrite? Should not you rewrite all the spatial functions?

Reply:
Unlike hash join where hash can be applied to multiple columns, spatial join works with a single relationship at a time. As I haven't seen queries with multiple spatial relations for a given pair of tables in production yet, I decided that support for multiple spatial relations in a single join is out of scope for this PR.